File: frontend/src/services/todoService.js
import axios from 'axios';

const API_BASE_URL = 'http://localhost:8080/api/todos';

const todoService = {
  // Get all todos
  getAllTodos: async () => {
    const response = await axios.get(API_BASE_URL);
    return response.data;
  },

  // Get todo by id
  getTodoById: async (id) => {
    const response = await axios.get(`${API_BASE_URL}/${id}`);
    return response.data;
  },

  // Create a new todo
  createTodo: async (todo) => {
    const response = await axios.post(API_BASE_URL, todo);
    return response.data;
  },

  // Update a todo
  updateTodo: async (id, todo) => {
    const response = await axios.put(`${API_BASE_URL}/${id}`, todo);
    return response.data;
  },

  // Delete a todo
  deleteTodo: async (id) => {
    await axios.delete(`${API_BASE_URL}/${id}`);
  },

  // Toggle completion status
  toggleComplete: async (id) => {
    const response = await axios.patch(`${API_BASE_URL}/${id}/toggle`);
    return response.data;
  },

  // Search todos
  searchTodos: async (searchTerm) => {
    const response = await axios.get(`${API_BASE_URL}/search`, {
      params: { q: searchTerm }
    });
    return response.data;
  },

  // Get todos by category
  getTodosByCategory: async (category) => {
    const response = await axios.get(`${API_BASE_URL}/category/${category}`);
    return response.data;
  },

  // Get todos by priority
  getTodosByPriority: async (priority) => {
    const response = await axios.get(`${API_BASE_URL}/priority/${priority}`);
    return response.data;
  },

  // Get overdue todos
  getOverdueTodos: async () => {
    const response = await axios.get(`${API_BASE_URL}/overdue`);
    return response.data;
  },

  // Get all categories
  getAllCategories: async () => {
    const response = await axios.get(`${API_BASE_URL}/categories`);
    return response.data;
  },

  // Get completed todos
  getCompletedTodos: async () => {
    const response = await axios.get(`${API_BASE_URL}/completed`);
    return response.data;
  },

  // Get incomplete todos
  getIncompleteTodos: async () => {
    const response = await axios.get(`${API_BASE_URL}/incomplete`);
    return response.data;
  }
};

export default todoService;

File: backend/src/main/java/com/todoapp/service/TodoService.java
package com.todoapp.service;

import com.todoapp.model.Todo;
import com.todoapp.repository.TodoRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class TodoService {
    
    @Autowired
    private TodoRepository todoRepository;
    
    // Get all todos
    public List<Todo> getAllTodos() {
        return todoRepository.findAll();
    }
    
    // Get todo by id
    public Optional<Todo> getTodoById(Long id) {
        return todoRepository.findById(id);
    }
    
    // Create a new todo
    public Todo createTodo(Todo todo) {
        return todoRepository.save(todo);
    }
    
    // Update an existing todo
    public Todo updateTodo(Long id, Todo todoDetails) {
        Todo todo = todoRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Todo not found with id: " + id));
        
        todo.setTitle(todoDetails.getTitle());
        todo.setDescription(todoDetails.getDescription());
        todo.setCompleted(todoDetails.isCompleted());
        todo.setPriority(todoDetails.getPriority());
        todo.setCategory(todoDetails.getCategory());
        todo.setDueDate(todoDetails.getDueDate());
        
        return todoRepository.save(todo);
    }
    
    // Delete a todo
    public void deleteTodo(Long id) {
        Todo todo = todoRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Todo not found with id: " + id));
        todoRepository.delete(todo);
    }
    
    // Toggle completion status
    public Todo toggleComplete(Long id) {
        Todo todo = todoRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Todo not found with id: " + id));
        
        todo.setCompleted(!todo.isCompleted());
        return todoRepository.save(todo);
    }
    
    // Get completed todos
    public List<Todo> getCompletedTodos() {
        return todoRepository.findByCompleted(true);
    }
    
    // Get incomplete todos
    public List<Todo> getIncompleteTodos() {
        return todoRepository.findByCompleted(false);
    }
    
    // SEARCH: Find todos by title
    public List<Todo> searchTodosByTitle(String searchTerm) {
        return todoRepository.findByTitleContainingIgnoreCase(searchTerm);
    }
    
    // Get todos by category
    public List<Todo> getTodosByCategory(String category) {
        return todoRepository.findByCategory(category);
    }
    
    // Get todos by priority
    public List<Todo> getTodosByPriority(Todo.Priority priority) {
        return todoRepository.findByPriority(priority);
    }
    
    // Get overdue todos
    public List<Todo> getOverdueTodos() {
        return todoRepository.findByDueDateBeforeAndCompletedFalse(LocalDate.now());
    }
    
    // Get all unique categories
    public List<String> getAllCategories() {
        return todoRepository.findAll().stream()
                .map(Todo::getCategory)
                .filter(category -> category != null && !category.isEmpty())
                .distinct()
                .sorted()
                .collect(Collectors.toList());
    }
}
